grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipImpl.java

```diff
@@ -118,6 +122,21 @@ public Relationship attributeRelationship(Attribute attribute) {
         return readFromReified((relationReified) -> relationReified.attributes(attributeTypes));
     }
 
+    @Override
+    public Stream<Attribute<?>> keys(AttributeType[] attributeTypes) {
+        Set<AttributeType> attributeTypesSet = Arrays.stream(attributeTypes).collect(Collectors.toSet());
+        Set<AttributeType> keyTypes = type().keys().collect(Collectors.toSet());
+
+        if(!attributeTypesSet.isEmpty()){
+            keyTypes = Sets.intersection(attributeTypesSet, keyTypes);
+        }
+
+        if(keyTypes.isEmpty()) return Stream.empty();
+
+        AttributeType[] keyTypesArray = keyTypes.toArray(new AttributeType[keyTypes.size()]);
+        return readFromReified((relationReified) -> relationReified.attributes(keyTypesArray));
+    }
```

Comment:
This is duplicated from `ThingImpl`. Can you instead delegate to the reified instance? If the relationship is not reified, then it has no keys... I think.